src/useScrollbarSize.ts

```diff
@@ -0,0 +1,59 @@
+import { useEffect, useRef, useState } from 'react';
+import { debounce } from './utils';
+
+export interface ScrollbarMeasurements {
+	/** Current height of the scrollbar */
+	height: number;
+	/** Current width of the scrollbar */
+	width: number;
+}
+
+const useScrollbarSize = (): ScrollbarMeasurements => {
+	const [height, setHeight] = useState(0);
+	const [width, setWidth] = useState(0);
+	const element = useRef<HTMLDivElement | null>(null);
+
+	// initialize resize event handler and state when mounted
+	useEffect(() => {
+		const getElement = () => {
+			if (element.current == null) {
+				// element was not created yet -- initialize
+				element.current = document.createElement('div');
+				element.current.style.width = '99px';
+				element.current.style.height = '99px';
+				element.current.style.overflow = 'scroll';
+				element.current.style.position = 'absolute';
+				element.current.style.top = '-9999px';
+				element.current.setAttribute('aria-hidden', 'true');
+				element.current.setAttribute('role', 'presentation');
+			}
+			return element.current;
+		};
+
+		const updateState = () => {
+			const { offsetHeight, clientHeight, offsetWidth, clientWidth } = getElement();
+			const scrollbarHeight = offsetHeight - clientHeight;
+			const scrollbarWidth = offsetWidth - clientWidth;
+			setHeight((h) => (h !== scrollbarHeight ? scrollbarHeight : h));
+			setWidth((w) => (w !== scrollbarWidth ? scrollbarWidth : w));
+		};
+
+		const handleResize = debounce(updateState, 100);
+
+		// initialize
+		window.addEventListener('resize', handleResize);
+		document.body.appendChild(getElement());
+		updateState();
+
+		// cleanup
+		return () => {
+			handleResize.cancel();
+			window.removeEventListener('resize', handleResize);
+			document.body.removeChild(getElement());
```

Comment:
Calling `getElement()` in the cleanup could be an issue since it would end up being a reference to `element.current`.  Usually, if you reference `myRef.current` in useEffect's cleanup you get a warning similar to this:

```
The ref value 'element.current' will likely have changed by the time this effect cleanup function runs. If this ref points to a node rendered by React, copy 'element.current' to a variable inside the effect, and use that variable in the cleanup function.
```

You may want to follow the warning's advice and copy `element.current` into a variable inside the effect and reference that variable in the cleanup function.  Something similar to this:

```javascript
const scrollElement = getElement();

// initialize
window.addEventListener('resize', handleResize);
document.body.appendChild(scrollElement);
updateState();

return () => {
  handleResize.cancel();
  window.removeEventListener('resize', handleResize);
  document.body.removeChild(scrollElement);
};
```

It's possible that this isn't an issue, but since it's something within the DOM, it may be best to play it safe.

Reply:
> warning about referencing a ref's value in the useEffect cleanup.

Thanks for such great comments, may you please share something about refrencing ref's value in useEffect cleanup. I had not heard of this till now. I did some searching and find only results that are refrencing the ref's value - https://medium.com/@alicantorun/react-hooks-usestate-and-useeffect-useref-with-cleanups-166d9231b6e4

Reply:
Oh nvm haha I just found a good issue. Its about refrencing dom elements that may not be there anymore! Thanks for this great comment! https://github.com/facebook/react/issues/15841

Reply:
Thank you.  I had originally gone with that approach, and had switched back once the element was being returned from the function.  However, you're probably right that this is a better safe than sorry scenario.